### PR TITLE
IPLD methods & testing

### LINKED ISSUE #88: Implement IPLD for public verifiable credentials storage
### Description

This task should implement the infrastructure needed to have a public credentials storage/service endpoint on the DDO.

IPLD will be used for easier routing to specific claims (tentatively, by claim ID).

### TODO
- [x] create DAG object from claim
- [ ] Public profile in data field of initial DAG object?
- [x] create DAG links between claims
- [x] methods to route to specific claims by claimID
- [ ] unit testing

### PR DESCRIPTION
**Pull request by charleenfei**
## Description

Fixes #88 .

> What does it solve? 

Implements IPLD add and retrieve DAG objects (CBOR encoding) methods & relevant tests.

> Were there any particular challenges that you faced and/or things that were not straightforward? 

## Checklist
- [x] Keep flags added to the PR up to date
- [x] Make sure the title describes the code well, description should contain `Fixes #number_of_an_issue` and a brief decscription of changes
- [x] Make sure the PR is readable and contains changes only related to a single issue. Otherwise - divide into separate PRs 
- [x] Add new tests and make sure all tests are passing
- [x] Make sure all edge cases are taken into account and covered by tests
- [x] Run linter and fix all the errors and warnings

### CODE REVIEW

File: tests/ipfs.test.ts [hunk 1]
```diff
@@ -0,0 +1,185 @@
...
+    nock('http://localhost:5001')
+    .post(`/api/v0/add?pin=${pinBoolean}`)
+    .reply(200, {
+      Name: 'testFile',
+      Hash: 'Qm12345',
+      Bytes: '12345',
+      Size: '12345'
+    })
+
+    try {
```
Comment (Exulansis) [inline tests/ipfs.test.ts]: Can this be changed to the `expect(...).to.throw('Error Message')`? Not sure how well that behaves with async / await code.

File: tests/ipfs.test.ts [hunk 2]
```diff
@@ -0,0 +1,185 @@
...
+
+  it('should attempt to retrieve the data', async () => {
+
+    nock('http://localhost:5001')
+    .get(`/api/v0/cat/${testHash}`)
+    .reply(200, {
+      file: 'test DDO file'
+    })
+
+    const resultingFile = await storageAgent.catJSON(testHash)
```
Comment (Exulansis) [inline tests/ipfs.test.ts]: Then section seems redundant.

File: tests/ipfs.test.ts [hunk 3]
```diff
@@ -0,0 +1,185 @@
...
+    .then((result) => {
+      return result
+    })
+    expect(resultingFile).to.deep.equal({file: 'test DDO file'})
```
Comment (Exulansis) [inline tests/ipfs.test.ts]:
```javascript
{file: 'test DDO file`}
```
could be a variable, so we don't have to maintain two string literals in sync.

File: tests/ipfs.test.ts [hunk 4]
```diff
@@ -0,0 +1,185 @@
...
+  })
+
+  it('should attempt to remove a pinned hash', async () => {
+
+    nock('http://localhost:5001')
```
Comment (Exulansis) [inline tests/ipfs.test.ts]: This URI can be a variable so we don't have to change it everywhere.

File: tests/ipfs.test.ts [hunk 5]
```diff
@@ -0,0 +1,185 @@
...
+    .get(`/api/v0/pin/rm?arg=${testHash}`)
+    .reply(200, {
+      Pins: [
+        `${testHash}`
+      ]
+    })
+
+    const pinRemoved = await storageAgent.removePinnedHash(testHash)
```
Comment (Exulansis) [inline tests/ipfs.test.ts]: Redundant `then` section?

File: tests/ipfs.test.ts [hunk 6]
```diff
@@ -0,0 +1,185 @@
...
+
+  it('should throw an error if removing the pinned hash failed', async () => {
+
+    nock('http://localhost:5001')
+    .get(`/api/v0/pin/rm?arg=${testHash}`)
+    .reply(400, {
+      error: 'Error'
+    })
+
+    try {
```
Comment (Exulansis) [inline tests/ipfs.test.ts]: Same comment with `expect.to.throw()`

File: tests/ipfs.test.ts [hunk 7]
```diff
@@ -0,0 +1,185 @@
...
+
+    const testDagData = { data: 'testData' }
+
+    nock('http://localhost:5001')
+    .post(`/api/v0/dag/put?pin=${pinBoolean}`)
+    .reply(200, {
+      Cid: { '/': testHash }
+    })
+
+    const resultingHash = await storageAgent.createDagObject(testDagData, pinBoolean)
```
Comment (Exulansis) [inline tests/ipfs.test.ts]: Same comment, then section.

File: tests/ipfs.test.ts [hunk 8]
```diff
@@ -0,0 +1,185 @@
...
+
+    const testDagData = 'testData'
+
+    nock('http://localhost:5001')
+    .post(`/api/v0/dag/put?pin=${pinBoolean}`)
+    .reply(200, {
+      Cid: { '/': testHash }
+    })
+
+    try {
```
Comment (Exulansis) [inline tests/ipfs.test.ts]: Same

File: ts/ipfs/index.ts [hunk 9]
```diff
@@ -40,5 +41,31 @@ export class IpfsStorageAgent implements IIpfsConnector {
     if (!res.ok) {
       throw new Error(`Removing pinned hash ${hash} failed, status code: ${res.status}`)
     }
+    return res.ok
```
Comment (Exulansis) [inline ts/ipfs/index.ts]: Should this return? Given that the return type is
```javascript
Promise<void>
```
Return type can be changed perhaps.

Comment (Exulansis) [reply ts/ipfs/index.ts]: Or simply don't return true, and if it does not throw, then it was successful

File: ts/ipfs/index.ts [hunk 10]
```diff
@@ -40,5 +41,31 @@ export class IpfsStorageAgent implements IIpfsConnector {
...
+  }
+
+  public async createDagObject(data: object, pin: boolean): Promise<string> {
+
+    if (typeof data !== 'object' || data === null) {
```
Comment (Exulansis) [inline ts/ipfs/index.ts]: Could be changed to 
```javascript
if (!data || typeof data !== 'object')
```
Because `data === null` will return `false` if the data is `undefined`
